### ISSUE TITLE
TypeError when calling isinstance on a TrameApp decorated object

### ISSUE THREAD

**Issue by MattTheCuber** [Kitware/trame#594]
**Describe the bug**

Using `isinstance` on a `@TrameApp()` decorated class throws a `TypeError`.

```
TypeError: isinstance() arg 2 must be a type, a tuple of types, or a union
```

**To Reproduce**

1. Run the following code.

***Code***

```python
from trame.decorators import TrameApp


@TrameApp()
class TestApp:
    server = None


t = TestApp()
print(isinstance(t, TestApp))
```

**Expected behavior**

I would assume `isinstance` would work. This is not a critical problem though as we have other workarounds.

The problem is that the `@TrameApp()` decorator returns a function that instantiates the class instead of returning a class itself. The `TypeError` is thrown because we are checking if a object is an instance of a function, which is not possible to check.

Maybe `isinstance` would never be a good way to handle this, either way it would be very useful to have a good way of checking the type of our app's instances.

**Platform:**

***OS:***

- [x] Windows
- [ ] MacOS
- [x] Linux
- [ ] Android
- [ ] iOS

**Comment by jourdain** [comment]
That is a good issue but I'm not sure how to solve that... maybe by doing the following

```python
from trame.decorators import TrameApp


class TestApp:
    server = None


t = TrameApp()(TestApp)()
print(isinstance(t, TestApp))
```

**Comment by MattTheCuber** [comment]
That is a solution, although fairly hard to follow in practice

**Comment by jourdain** [comment]
yes, I agree and I'm not sure how to solve that.

**Comment by MattTheCuber** [comment]
I asked Chat and it suggested creating a wrapper class with the decorator.

```python
class TrameApp:
    def __call__(self, klass):
        class WrappedClass(klass):
            ...

        return WrappedClass
```

It could work, but then it could introduce other problems with type checking since `type(t)` would return `WrappedClass` instead of `TestApp`.

**Comment by psavery** [comment]
We might be able to re-design the class a little bit to support this. I might look into it later this week.

**Comment by MattTheCuber** [comment]
I just found another issue that the function wrapper may be causing. Not quite sure why this is occuring though:

```python
from trame.decorators import TrameApp


@TrameApp()
class TestApp:
    server = None


@TrameApp()
class TestApp2(TestApp):
    server = None


t = TestApp()
t2 = TestApp2()
```

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[2], line 10
      4 @TrameApp()
      5 class TestApp:
      6     server = None
      9 @TrameApp()
---> 10 class TestApp2(TestApp):
     11     server = None
     14 t = TestApp()

TypeError: function() argument 'code' must be code, not str
```

**Comment by jourdain** [comment]
I think this is related because the decorated TestApp is actually a function and not a class which prevent you from inheriting it.

**Comment by MattTheCuber** [comment]
Hey @psavery, I just wanted to check in on the status of this?

**Comment by MattTheCuber** [comment]
I believe both of these issues are fixed now.

**Comment by jourdain** [comment]
The latest trame also has a class that you can inherit

**Comment by jourdain** [comment]
And all the widgets now inherit TrameComponent to support the class decorators. So no need to decorate a widget anymore.
